### ISSUE TITLE
Processes from different groups are receiving all messages from all groups

### ISSUE THREAD

**Issue by pyke369** [Exa-Networks/exabgp#268]
Hello @thomas-mangin,

There seems to be a problem when trying to declare multiple groups with different BGP peers and processes definition: all processes are receiving all messages from all groups. 

Let's say you have the following ExaBGP configuration:

```
group group1 {
  process external1 {
    encoder json;
    peer-updates;
    neighbor-changes;
    receive-routes;
    run external.py;
  }
  neighbor 10.140.0.1 {
    router-id 192.168.37.128;
    local-address 192.168.37.128;
    local-as 65501;
    peer-as 65500;
    family {
      inet4 unicast;
      inet6 unicast;
    }
  }
}
group group2 {
  process external2 {
    encoder json;
    peer-updates;
    neighbor-changes;
    receive-routes;
    run external.py;
  }
  neighbor 10.140.0.2 {
    router-id 192.168.37.128;
    local-address 192.168.37.128;
    local-as 65501;
    peer-as 65500;
    family {
      inet4 unicast;
      inet6 unicast;
    }
  }
}
```

The external.py script is just dumping everything it receives in a file:

```
#!/usr/bin/env python

import sys, select, os

handle = open('/tmp/x%d.log' % os.getpid(), 'a+');
while True:
    ready, _, _ = select.select([sys.stdin], [], [], 1)
    if ready:
        line = sys.stdin.readline().strip()
        handle.write(line + "\n")
        handle.flush()
```

When starting the daemon (v3.4.10), I get this:

```
$ exabgp external.conf
\[...\]
Sun, 17 May 2015 19:49:32 | INFO     | 29945  | reactor       | New peer setup: neighbor 10.140.0.2 local-ip 192.168.37.128 local-as 65501 peer-as 65500 router-id 192.168.37.128 family-allowed in-open
Sun, 17 May 2015 19:49:32 | INFO     | 29945  | reactor       | New peer setup: neighbor 10.140.0.1 local-ip 192.168.37.128 local-as 65501 peer-as 65500 router-id 192.168.37.128 family-allowed in-open
Sun, 17 May 2015 19:49:32 | WARNING  | 29945  | configuration | Loaded new configuration successfully
Sun, 17 May 2015 19:49:32 | INFO     | 29945  | processes     | Forked process external2
Sun, 17 May 2015 19:49:32 | INFO     | 29945  | processes     | Forked process external1
Sun, 17 May 2015 19:49:35 | INFO     | 29945  | network       | Connected to peer neighbor 10.140.0.1 local-ip 192.168.37.128 local-as 65501 peer-as 65500 router-id 192.168.37.128 family-allowed in-open (out)
Sun, 17 May 2015 19:49:35 | INFO     | 29945  | network       | Connected to peer neighbor 10.140.0.2 local-ip 192.168.37.128 local-as 65501 peer-as 65500 router-id 192.168.37.128 family-allowed in-open (out)
^CSun, 17 May 2015 19:49:53 | INFO     | 29945  | reactor       | ^C received
Sun, 17 May 2015 19:49:53 | INFO     | 29945  | reactor       | Performing shutdown
Sun, 17 May 2015 19:49:54 | INFO     | 29945  | processes     | Terminating process external2
Sun, 17 May 2015 19:49:54 | INFO     | 29945  | processes     | Terminating process external1
```

Both resulting files contain the exact same messages:

```
$ ls -al /tmp/x*.log
-rw-r----- 1 root root 4089 mai   17 19:49 /tmp/x29948.log
-rw-r----- 1 root root 4089 mai   17 19:49 /tmp/x29949.log

$ diff -s  /tmp/x*.log
Les fichiers /tmp/x29948.log et /tmp/x29949.log sont identiques
```

Is there a way to segregate messages for each group and send them only to the associated process? I saw a message referring to "group-updates not being enabled by default" in ExaBGP log but I couldn't find any documentation. The exact lines were:

```
group-updates not enabled for peer 10.140.0.1, it surely should, the default will change to true soon
group-updates not enabled for peer 10.140.0.2, it surely should, the default will change to true soon
``` 

(I'm not sure it's even relevant to that particular problem...)

Thanks,
Pierre-Yves

**Comment by thomas-mangin** [comment]
This is a issue with the current configuration and internals which I want to fix in the next configuration format ( by explicitly assigning the process to the neighbors)

The simplest way to get around it is to create two configuration file and run exabgp with both in the command line. It will fork two process each handling one configuration file.

As it will not be a small fix, I will do it during before the release of 3.5.0 but I will not back port it to 3.4 because it will change the behaviour of the current installation.

**Comment by pyke369** [comment]
I did some additional testing, and it seems putting the processes definition inside the neighbors sections lead to the correct behavior: each process only receives the messages from the associated neighbor and the associated neighbor only. The problem is that I'd like to have multiple neighbors within a group sharing a single external process.

I think the correct behavior should be that an associated group-level process would only receive messages sent by all neighbors within that group, and send any messages generated by that process to all neighbors within the group.

Starting several instances of the daemon is not really feasible in my current configuration, but I may find a way to filter messages out, based on the group each neighbor is in (another simple way would be to augment the messages syntax with a "group": "name" field).

I'll let you know what I'll come up with...

**Comment by thomas-mangin** [comment]
I agree with you, it is just not simple/practical to implement with the code as-it.

**Comment by thomas-mangin** [comment]
As this is not a bug per say but a bad design of the configuration parser, will be addressed in #280
